### PR TITLE
Add support for TDS v2 (relation) groupBy in QueryBuilder

### PR DESCRIPTION
**Pull request by travisstebbins**
## Summary

Add support for using the new TDS v2 (relation) protocol with groupBy expressions. This includes
 - Supporting converting the QueryBuilder state to the new protocol when the user has `groupBy` functions (i.e., count, sum, etc.)
 - Supporting converting the new protocol with `groupBy` functions to QueryBuilder state

## How did you test this change?

- [x] Test(s) added
- [ ] Manual testing (please provide screenshots/recordings)
- [ ] No testing (please provide an explanation)

Group by:
GroupByQuery

Group by with multiple columns:
MultiGroupByQuery

Group by with pre-filter:
PreFilter

Group by with post-filter:
PostFilter

Group by with pre-filter and post-filter:
PreFilterAndPostFilter

### CODE REVIEW

File: packages/legend-query-builder/src/stores/QueryBuilderStateBuilder.ts [hunk 1]
```diff
@@ -740,16 +745,24 @@ export class QueryBuilderValueSpecificationProcessor
...
-      )
+        QUERY_BUILDER_SUPPORTED_FUNCTIONS.RELATION_GROUP_BY,
+      ])
     ) {
-      processTDSGroupByExpression(
-        valueSpecification,
-        this.queryBuilderState,
-        this.parentLambda,
-      );
+      if (isTypedGroupByExpression(valueSpecification)) {
```
Comment (akphi) [inline packages/legend-query-builder/src/stores/QueryBuilderStateBuilder.ts:753]: same comment here, we should be careful of the case where people specify the fully specified function path

**Review by akphi — COMMENTED** [review]
Nice job! The tests and general flow are very comprehensive.

I left some minor comments, mostly on some edge cases handling and code organization.

File: .changeset/light-icons-sparkle.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+---
+'@finos/legend-query-builder': patch
+'@finos/legend-graph': patch
+---
+
+Add support for TDS v2 (relation) groupBy in QueryBuilder
```
[the remark this answers is not in the document]
Comment (travisstebbins) [reply .changeset/light-icons-sparkle.md:5]: Done

File: packages/legend-query-builder/src/graph-manager/helpers/QueryBuilderAggregateOperatorHelper.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+ */
+
+import {
+  type Type,
+  matchFunctionName,
+  PrimitiveType,
+} from '@finos/legend-graph';
+import { QUERY_BUILDER_SUPPORTED_FUNCTIONS } from '../../graph/QueryBuilderMetaModelConst.js';
+
+export const getNumericAggregateOperatorReturnType = (
```
[the remark this answers is not in the document]
Comment (travisstebbins) [reply packages/legend-query-builder/src/graph-manager/helpers/QueryBuilderAggregateOperatorHelper.ts]: Makes sense! I removed the use of this function from all the operators. I made it a non-exported function in `packages/legend-query-builder/src/graph-manager/protocol/pure/v1/V1_QueryValueSpecificationBuilderHelper.ts` since that's the only place it's used now.

File: packages/legend-query-builder/src/graph-manager/protocol/pure/v1/V1_QueryValueSpecificationBuilderHelper.ts [hunk 4]
```diff
@@ -1140,6 +1141,226 @@ export const V1_buildGroupByFunctionExpression = (
...
+  ).accept_ValueSpecificationVisitor(
+    new V1_ValueSpecificationBuilder(
+      compileContext,
+      processingContext,
+      openVariables,
+    ),
+  );
+  assertNonNullable(
+    precedingExpression.genericType,
+    `Can't build groupBy() expression: preceding expression return type is missing`,
```
[the remark this answers is not in the document]
Comment (travisstebbins) [reply packages/legend-query-builder/src/graph-manager/protocol/pure/v1/V1_QueryValueSpecificationBuilderHelper.ts]: Done

File: packages/legend-query-builder/src/stores/fetch-structure/tds/aggregation/operators/QueryBuilderAggregateOperator_Average.ts [hunk 5]
```diff
@@ -96,7 +100,11 @@ export class QueryBuilderAggregateOperator_Average
   override getReturnType(
     aggregateColumnState: QueryBuilderAggregateColumnState,
   ): Type {
-    return PrimitiveType.FLOAT;
+    return guaranteeNonNullable(
+      getNumericAggregateOperatorReturnType(
```
[the remark this answers is not in the document]
Comment (travisstebbins) [reply packages/legend-query-builder/src/stores/fetch-structure/tds/aggregation/operators/QueryBuilderAggregateOperator_Average.ts]: Done

**Review by akphi — APPROVED** [review]
